### PR TITLE
Merged getting all contributors and recent contributors

### PR DESCRIPTION
**Pull request by szczepiq**
Merged getting all contributors and recent contributors. In the process, I've cleaned up a lot of code :) The change is massive, sorry about that. Reviewing commit-by-commit may not be helpful because I have changed my mind about implementation during the development of this.

For motivation, see #120

### CODE REVIEW

**Comment by szczepiq** [comment]
Any feedback? Give it at least a short pass / eye-ball the changes. Perhaps some of the implementation will be interesting to you.

File: src/test/groovy/org/mockito/release/internal/gradle/util/pom/PomXmlCustomizerTest.groovy [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package org.mockito.release.internal.gradle.util.pom
+
+import spock.lang.Specification
+
+class PomXmlCustomizerTest extends Specification {
+
```
Comment (mstachniuk) [inline src/test/groovy/org/mockito/release/internal/gradle/util/pom/PomXmlCustomizerTest.groovy]: Not needed.

File: src/main/groovy/org/mockito/release/internal/gradle/util/PomCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                                + StringUtil.join(conf.getTeam().getContributors(), ", "));
+                customizePom(xml.asNode(), conf, archivesBaseName, project.getDescription());
+            }
+        });
+    }
+
+    /**
+     * Customizes pom xml based on the provide configuration and settings
+     */
+    public static void customizePom(Node root, ReleaseConfiguration conf,
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/internal/gradle/util/PomCustomizer.java]: +1

File: src/main/groovy/org/mockito/release/internal/gradle/util/PomCustomizer.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                d.appendNode("url", "https://github.com/" + person.gitHubUser);
+            }
+        }
+
+        if (!conf.getTeam().getContributors().isEmpty()) {
+            Set<String> devs = new HashSet<String>(conf.getTeam().getDevelopers());
+            Node contributors = root.appendNode("contributors");
+            for (String notation : conf.getTeam().getContributors()) {
+                if (!devs.contains(notation)) {
+                    ReleaseConfigurationTeamParser.Person person = parsePerson(notation);
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/internal/gradle/util/PomCustomizer.java]: Nice feedback, +1

File: src/main/groovy/org/mockito/release/internal/gradle/util/PomCustomizer.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                    Node d = contributors.appendNode("contributor");
+                    d.appendNode("name", person.name);
+                    d.appendNode("url", "https://github.com/" + person.gitHubUser);
+                }
+            }
+        }
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/internal/gradle/util/PomCustomizer.java]: This is done, there is a "configureContributors" task that configures 'releasing.team.contributors' collection. Unless I'm missing something :)

File: src/main/groovy/org/mockito/release/notes/contributors/RecentContributorsFetcher.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.*;
+
+import static org.mockito.release.internal.util.ArgumentValidation.notNull;
+import static org.mockito.release.internal.util.DateUtil.forGitHub;
+import static org.mockito.release.internal.util.DateUtil.yesterday;
+
+/**
+ * Fetches recent contributors from GitHub using the "commit" end point.
+ * The "contributors" endpoint does not return the most recent contributors as documented.
+ */
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/notes/contributors/RecentContributorsFetcher.java]: +1, Done

File: src/main/groovy/org/mockito/release/notes/contributors/RecentContributorsFetcher.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+public class RecentContributorsFetcher {
+
+    private static final Logger LOG = Logging.getLogger(RecentContributorsFetcher.class);
+
+    /**
+     * Contributors that pushed commits to the repo withing the last 24hrs
+     */
+    public Collection<Contributor> fetchContributorsSinceYesterday(String repository, String readOnlyAuthToken) {
+        String yesterday = forGitHub(yesterday());
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/notes/contributors/RecentContributorsFetcher.java]: 24h worked for me :) We'll see if notes have right stuff. We can always come back and make it 48h. GitHub docs say contributors endpoint is 4h delayed so 24 should be ok.

File: src/main/groovy/org/mockito/release/notes/contributors/RecentContributorsFetcher.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    /**
+     * Contributors that pushed commits to the repo within the time span.
+     * @param dateSince - must not be null, the since date, in format accepted by GitHub
+     *                  ({@link org.mockito.release.internal.util.DateUtil#forGitHub(Date)})
+     * @param dateUntil - can be null, it means there is no end date, in format accepted by GitHub
+     *                  ({@link org.mockito.release.internal.util.DateUtil#forGitHub(Date)})
+     */
+    public Collection<Contributor> fetchContributors(String repository, String readOnlyAuthToken, String dateSince, String dateUntil) {
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/notes/contributors/RecentContributorsFetcher.java]: Nice feedback, thank you. Done.

File: src/main/groovy/org/mockito/release/notes/contributors/RecentContributorsFetcher.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+                this.dateUntil = dateUntil;
+            }
+
+            GitHubCommits build() {
+                // see API doc: https://developer.github.com/v3/repos/commits/#list-commits-on-a-repository
+                String nextPageUrl = "https://api.github.com/repos/" + repository + "/commits"
+                        + "?access_token=" + readOnlyAuthToken
+                        + "&since=" + dateSince
+                        + ((dateUntil != null)? "&until=" + dateUntil : "")
+                        + "&page=1&per_page=100";
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/notes/contributors/RecentContributorsFetcher.java]: Considered, tried, but didn't seem more readable. Left it as is. I don't have strong opinion - if you wander in this part of the code and want to change it, go for it!

File: src/main/groovy/org/mockito/release/notes/contributors/GitHubCommitsJSON.java [hunk 9]
```diff
@@ -20,13 +20,8 @@ static Contributor toContributor(JsonObject commit) {
             String profileUrl = (String) ((Map) commit.get("author")).get("html_url");
             return new DefaultContributor(name, login, profileUrl);
         } catch (NullPointerException e) {
+            //TODO when does it happen?
             return null;
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/notes/contributors/GitHubCommitsJSON.java]: Nice! Thank you.

File: src/test/groovy/org/mockito/release/internal/gradle/ReleaseConfigurationTest.groovy [hunk 10]
```diff
@@ -21,4 +27,22 @@ class ReleaseConfigurationTest extends Specification {
         conf.build.commitMessage == null
         !conf.build.pullRequest
     }
+
+    def "validates team members"() {
```
Comment (szczepiq) [inline src/test/groovy/org/mockito/release/internal/gradle/ReleaseConfigurationTest.groovy]: Great catch. This test was not working :) I updated it.

File: src/test/groovy/org/mockito/release/internal/gradle/util/pom/PomXmlCustomizerTest.groovy (hunk 1 above)
Comment (szczepiq) [inline src/test/groovy/org/mockito/release/internal/gradle/util/pom/PomXmlCustomizerTest.groovy]: Ha, dead code. Well spotted.

**Comment by szczepiq** [comment]
I pushed the fixes and when the build is green, I'm merging!

File: src/main/groovy/org/mockito/release/internal/gradle/util/PomCustomizer.java (hunk 4 above)
Comment (mstachniuk) [inline src/main/groovy/org/mockito/release/internal/gradle/util/PomCustomizer.java]: Ohh ok, I see now. I thought `conf.getTeam().getContributors()` returns only contributors defined in gradle/release.gradle file, but now I see code in `ConfigureContributorsTask`:

```
releaseConfiguration.getTeam().setContributors(contributors);
```

So general design question: Can we use `ReleaseConfiguration` as global state to exchange data between tasks like here?

File: src/main/groovy/org/mockito/release/notes/contributors/RecentContributorsFetcher.java (hunk 8 above)
Comment (mstachniuk) [inline src/main/groovy/org/mockito/release/notes/contributors/RecentContributorsFetcher.java]: Readability FTW

**Pull request merged**